src/main/java/homework_4/custom_file_reader/CustomFileReader.java

```diff
@@ -0,0 +1,57 @@
+package homework_4.custom_file_reader;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Scanner;
+
+import static java.lang.System.lineSeparator;
+
+public class CustomFileReader {
+    private static final String FILE_NAME = "src/main/resources/custom_file_reader/Shakespeare.txt";
+    private static final String ERROR_MESSAGE = "Can not read file.";
+    private static final File FILE = new File(FILE_NAME);
```

Comment:
do you really need this File creating even without CustomFileReader object?
I see no static methods to operate with it.

As for me, it's better to create File instances in methods working with it. At least you could catch and process Exceptions, e.g. FileNotFoundException.

Reply:
Thank you for the rewiev. When I did this assignment I assumed that:
1. File has private access, it cannot be used outside of class instances.

2. Constants are some constant values ​​for all instances of a class, and according to the conditions of the homework assignment, we will have exactly one file. Therefore, taking into account that the File itself is immutable, and good practice is to use constant values ​​where we know that they should not change, I made a constant File variable to prevent unnecessary creation of this object for different instances since our file is always the same.

3. As far as I looked, the File constructor does not throw an exception, there goes a file name normalization using standard methods and the resource does not open at this stage.
So all the work on opening the resource and working directly with the file will take place in the methods.

For example, FileNotFoundException will be handled by the catch (IOException e) block of the corresponding method.

This is what I was guided by, I understand that it was possible and so, but I relied on my understanding and the condition of the assignment, perhaps there are some drawbacks of this approach, but I do not see them yet, and I am not sure within the framework of this specific assignment are whether they are significant considering the YAGNI principle

Reply:
I get you. 
In my opinion, according to OOP paradigm it should look like in a little bit different way:
First of all, constuctor could be created, where you assign values to your class fields.
Then Your App is operating with a CustomFileReader object, where File object is assigned to the CustomFileReader object itself, not to the class. Thus, File object **exists only** if an CustomFileReader object created.

_Why `private static final File FILE = new File(FILE_NAME);` is bad practice?_
 _Consider you need to read N different files with your App. Yeah, your task has changed suddenly :) BUT! You already wrote the program to solve almost the same.
In the way i described above you should already have an empty constuctor and you should just add one more with parameter like CustomFileReader(String file_Name) {....} (or create setter for file_Name).
Anyway, from here you'll be able to work with File object assigned to your CustomFileReader object. No more files loaded - no more File objects created. The only one you need to operate with.
Your architecture is unsuitable for solving this task in the current form._

P.S.
I should say that you solved the specific task more efficient then if you use my "advice", but this way is not flexible at all.

Reply:
I get your points, but again, the same task could be done in many ways, this is just my suggestion.

Java is not pure OOP language due to language design with static, primitives, and other features of structure-procedural languages, therefore, the line between pure OOP and procedural-structural ways is pretty thin. For example, System.in and System.out variables are declared as final static, and JVM initializes it with console input/output resources at each start of JVM. That's only one from many examples, how Java violates pure OOP principles, which abides only Smalltalk and maybe a few other very specific languages. All other OOP languages tend to seek a balance between the OOP paradigm and functionality. As it seems to me, Java combines best from OOP and structural-procedural approaches and provides a high level of agile and scalable solutions for complex tasks.

About the flexibility of this particular code, again, what we are talking about is a balance between reasonable premature optimization and the YAGNI principle. But the point is, every solution has its pro and cons, I fully agree with you in terms that my solution is not as flexible as it could be.

As I assume, when we are talking about developing complex solutions for customers, we should always keep it in mind, that our code design and architecture should be flexible enough to be able to provide agile refactoring and extending of functionality, but at the same time we should take in account that a customer pays not for the features that he might not even need, but for the specific feature here and now, and we should have a sufficient reason to think, that efforts spent on premature optimization will be justified in future.

Everything should be balanced between solving a specific task with not flexible solutions from one side and overengineering from the other side. This particular solution is just my bet on a specific case.

I agree with your words and admit that this solution is just a specific solution for a specific task and not flexible for other possible tasks though I guess it won't take much time to refactor this code to new conditions since it's really few lines. But at the same time, I think that your point of view is valid, but still, our mistakes are our experience :)

It's almost impossible to say, what is the best option, we are using our sense, which is based on our experience. An important way of developing this sense to choose what tool or approach should be applied to a task is to make an assumption and try it :) And when we have got some results we can analyze what worked well and what did not.

I am very grateful for your feedback and discussion. That's how developing skills become better. I rethought and revised many things. But I just train this balance, that I mentioned above, and want to explore and understand the borders of an acceptable solution.